smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/tagging/TaggableResourceApiValidator.java

```diff
@@ -0,0 +1,207 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.PropertyBindingIndex;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates service satisfies AWS tagging requirements.
+ */
+public final class TaggableResourceApiValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new LinkedList<>();
+        AwsTagIndex tagIndex = AwsTagIndex.of(model);
+        PropertyBindingIndex propertyBindingIndex = PropertyBindingIndex.of(model);
+        for (ServiceShape service : model.getServiceShapesWithTrait(TagEnabledTrait.class)) {
+            for (ShapeId resourceId : service.getResources()) {
+                ResourceShape resource = model.expectShape(resourceId).asResourceShape().get();
+                events.addAll(validateResource(model, resource, service, tagIndex, propertyBindingIndex));
+            }
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> validateResource(
+            Model model,
+            ResourceShape resource,
+            ServiceShape service,
+            AwsTagIndex awsTagIndex,
+            PropertyBindingIndex propertyBindingIndex
+    ) {
+        List<ValidationEvent> events = new LinkedList<>();
+        // Generate warning if resource has tag property in update API.
+        if (awsTagIndex.isResourceTagOnUpdate(resource.getId())) {
+            Shape updateOperation = model.expectShape(resource.getUpdate().get());
+            events.add(danger(updateOperation, "Update resource lifecycle operation should not support updating tags "
+                    + " because it is a privileged operation that modifies access."));
+        }
+        // A valid taggable resource must support one of the following:
+        // 1. Tagging via service-wide TagResource/UntagResource/ListTagsForResource
+        //    APIs that have input which targets a resource ARN and have expected tag
+        //    list or tag keys input or output member
+        // 2. Tagging via APIs specified in the @taggable trait which are validated
+        //    through the tag property, and must be resource instance operations
+        // Note: Cannot mix and match #2 and #3 within the same service but a resource
+        //    can support all three.
+        boolean isTaggable = false;
```

Comment:
Yes, but because `isTaggableViaInstanceOperations()` must be run to generate validation events even if taggable via service API's is also true. Having a single `||` makes me worry that someone may come through and swap the order. Though a unit test should catch this, it would still cost a dev some time to go back and understand how this caused it.

Reply:
Maybe that should be refactored to a `get` that returns a list of events and can be checked via `isEmpty()`?

Reply:
Maybe, but then it runs the risk of that encapsulated validation having events that don't change the taggable determination.